### ISSUE TITLE
[no-unused-vars] Wrong position reported

### ISSUE THREAD

**Issue by bachorp** [typescript-eslint/typescript-eslint#4910]
**Repro**

```JSON
{
    "parser": "@typescript-eslint/parser",
    "plugins": [
        "@typescript-eslint"
    ],
    "rules": {
        "@typescript-eslint/no-unused-vars":
            "error"
    }
}
```

```TS
function foo(x: null) {
    return x
}

let x = null

x = foo(x)
```

**Expected Result**

```
7:1  error  'x' is assigned a value but never used.  @typescript-eslint/no-unused-vars

x = foo(x)
^ error here
```

**Actual Result**

```
7:9  error  'x' is assigned a value but never used.  @typescript-eslint/no-unused-vars

x = foo(x)
        ^ error here
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.22.0` |
| `@typescript-eslint/parser`        | `5.22.0` |
| `TypeScript`                       | `4.6.2` |
| `ESLint`                           | `8.14.0` |
| `node`                             | `16.12.0` |

**Comment by bradzacher** [comment]
Could you more clearly annotate which position you expect and which position you got?

You said you expect line 1 column 1 which is clearly a typo because that would suggest you want the `function` keyword underlined.

**Comment by bachorp** [comment]
@bradzacher Sorry, I updated my comment

**Comment by bradzacher** [comment]
@bachorp theres only 7 lines in your example. So `9:1` doesn't make much more sense.

As per my last comment (I edited after you saw it I think) could you annotate your example code with comments showing the actual and expected locations?

**Comment by bachorp** [comment]
@bradzacher Done

**Comment by bradzacher** [comment]
The rule is actually right here

```ts
let x = null
``` 

In this code the variable `x` is not used.

```ts
let x = null
foo(x)
```

In this code the variable `x` is used after it is assigned a value.

```ts
let x = null
foo(x)
x = 1 // error on this line
```

In this code you will have an error as the last line assigns a value to `x` which is never used.

Putting these together:

```ts
    let x = null

    x = foo(x)
//          ^ usage of the declaration ✅
//  ^ new assignment to x which is never used  ❌
```

**Comment by bachorp** [comment]
@bradzacher Can you please clarify at which position we would expect the error? If I understand your example correctly (we expect an error at ❌  but not at ✅), the rule is **not** right here.

**Comment by bradzacher** [comment]
Ah - I see the bug now.
Yes, our rule is showing the wrong error and does not match the base rule

https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=PQKgBApgzgNglgOwC5gQewLQFcFahAEwwDcBDAJygC5Jzy1ywRgAoUSWRFAASQE8ADtADG5OAKQZo8ZMHTZc+ImUo0IdBk1YsAZjmFI4aBGB1o0ACgAeASjABvFmGdhyEJFnImrLAL4sWGHcwKzAAXlQsGBgA0IizS1sgA&rules=N4XyA&tsConfig=N4XyA

**Comment by armano2** [comment]
there actually seem to be additional issues, and some of test seem to be incorrect,
```ts
interface Foo {
  bar: string;
  baz: Foo['bar'];
  //   ^ reports this
}
```

**Comment by bradzacher** [comment]
@armano2 - that's the same issue - we're reporting the last read instead of the last write.

I don't think we ever merged this from upstream: https://github.com/eslint/eslint/commit/5df5e4a9976964fcf4dc67e241d4e22ec1370fe0

Separately there are the following commits we should assess and merge in as well:
- https://github.com/eslint/eslint/commit/9e9b5e07475564813b62cd1d7562a93c5fb4bc74
- https://github.com/eslint/eslint/commit/f110926a7abcc875a86dd13116f794e4f950e2ba
- https://github.com/eslint/eslint/commit/800bd258e4484de24323809ebbf13fc72fcbabac
- https://github.com/eslint/eslint/commit/f364d47013d146cdea42b27a7b24d105223f5ffe

**Comment by armano2** [comment]
correct, we do not have those changes and after testing our rule is not reporting on them correctly, do you mind filling separate issue about this?

**Comment by bradzacher** [comment]
I also didn't realise that this was a duplicate of #3311 oops
